diff --git a/libopenconnect.map.in b/libopenconnect.map.in
index f036c07..2ca187b 100644
--- a/libopenconnect.map.in
+++ b/libopenconnect.map.in
@@ -131,2 +131,3 @@ OPENCONNECT_5_9 {
  global:
+	openconnect_get_tun_fd;
 	openconnect_set_sni;
diff --git a/libopenconnect5.symbols b/libopenconnect5.symbols
index 7cb04d0..cd54f80 100644
--- a/libopenconnect5.symbols
+++ b/libopenconnect5.symbols
@@ -103,2 +103,3 @@ libopenconnect.so.5 libopenconnect5 #MINVER#
  openconnect_set_mca_key_password@OPENCONNECT_5_8 9.00
+ openconnect_get_tun_fd@OPENCONNECT_5_9 9.12
  openconnect_set_useragent@OPENCONNECT_5_8 9.00
diff --git a/library.c b/library.c
index 840846d..786cff5 100644
--- a/library.c
+++ b/library.c
@@ -1593,4 +1593,9 @@ int openconnect_setup_tun_device(struct openconnect_info *vpninfo,
 }
 
+int openconnect_get_tun_fd(struct openconnect_info *vpninfo)
+{
+	return vpninfo->tun_fd;
+}
+
 static const char * const compr_name_map[] = {
 	[COMPR_DEFLATE] = "Deflate",
diff --git a/openconnect.h b/openconnect.h
index 5160937..7865b98 100644
--- a/openconnect.h
+++ b/openconnect.h
@@ -835,4 +835,6 @@ typedef void (*openconnect_setup_tun_vfn) (void *privdata);
 void openconnect_set_setup_tun_handler(struct openconnect_info *vpninfo,
 				       openconnect_setup_tun_vfn setup_tun);
+int openconnect_get_tun_fd(struct openconnect_info *vpninfo);
+
 
 /* Callback for indicating that a TCP reconnection succeeded. */
